docs: Drafted Software Distribution Hub

ashokdevtron opened the pull request:
- Created dedicated section (README.md, tenants.md, release-hub.md)
- Created an entry in SUMMARY.md
- Showed tenants as a prerequisite to releasehub-related actions
- Added Enterprise tagging
- Added visuals (snapshots, gifs, videos)
- Explained the purpose in README.md (Why and when and who)
- Gave examples wherever necessary for the user-input fields
- Explained rescind and on-hold actions in Extras
- Showed Partial Release and Full Release in Extras
- Included RBAC (super-admin as of now)

abhibhaw commented on docs/user-guide/sdh/README.md:
```diff
@@ -0,0 +1,79 @@
+
+* **Standardized Deployment Processes**: You want to standardize and automate deployment processes to ensure error-free releases and streamline operations across multiple tenant environments.
+
+* **Comprehensive Release Documentation**: You require detailed documentation and checklists for each release, specifying versioning, dependencies, configuration changes, and deployment prerequisites to ensure consistent and successful deployments.
```
This area can be better reframed from problems lens. Currently its from solution lens

abhibhaw commented on docs/user-guide/sdh/README.md:
```diff
@@ -0,0 +1,79 @@
+| **Release Management**                     | No versioned deployments                     | Centralizes versioning and deployment into a unified platform  |
+| **Visibility**                             | Limited visibility (Apps/App Groups)         | Comprehensive visibility                                       |
+| **Automated and Standardized Deployments** | Relies on manual scripts or basic automation (jobs) | Automates and standardizes deployments for consistency  |
+| **End-to-End Release Management**          | Focuses on pushing changes quickly           | Manages entire release lifecycle from planning to post-release |
```
collaboration can be an aspect to consider here

abhibhaw commented on docs/user-guide/sdh/README.md:
```diff
@@ -0,0 +1,79 @@
+
+Feel free to familiarize yourself with the following concepts (terms) before you proceed to Software Distribution Hub.
+
+### Tenants
```
Are we planning to add the new entities to glossary?

ashokdevtron replied:
No, @abhibhaw
These seem to be SDH-specific terms at the moment, so we intend to keep all these terms in this doc itself. In future, if we reuse any of these terms outside the SDH doc, we will put it in the glossary. Abhi ke liye, it is a one-stop vocabulary section for educating our users upfront.

abhibhaw commented on docs/user-guide/sdh/README.md:
```diff
@@ -0,0 +1,79 @@
+
+One installation is equivalent to one deployment of your software for a specific [tenant](#tenants). Each installation is a different instance of the software tailored to different stages of use. For example, Microsoft might have three installations: 
+
+* **Production Installation**: Where the live environment of the software runs for everyday use.
```
Alternate way of saying this could be team based instances, seperate instance for different teams of an organisations

abhibhaw commented on docs/user-guide/sdh/README.md:
```diff
@@ -0,0 +1,79 @@
+
+### Release Order/Stage
+
+This is a part of [requirements](#requirements) where you decide the stages in which applications are deployed. This ensures that all dependencies are met and the software is implemented correctly. For example, you might need to deploy a database update before deploying a new application feature that relies on that update. In such a case, release order ensures that the backend applications are de ...
```
~For example, you might need to deploy a database update before deploying a new application feature that relies on that update.~ Doesn't seem very connected with the ending statement, database updates are generally migrations, which are incorporated with jobs generally, which shouldn't be handled with release stages

ashokdevtron replied:
Fixed

abhibhaw commented on docs/user-guide/sdh/README.md:
```diff
@@ -0,0 +1,79 @@
+
+This is a part of [requirements](#requirements) where you decide the stages in which applications are deployed. This ensures that all dependencies are met and the software is implemented correctly. For example, you might need to deploy a database update before deploying a new application feature that relies on that update. In such a case, release order ensures that the backend applications are de ...
+
+### Rollout
```
I'm a not very sure on the description provided, rolling out a release is verb?

ashokdevtron replied:
Yes, 'rolling out a release' is verb
But now I rephrased the whole thing...

abhibhaw commented on docs/user-guide/sdh/release-hub.md:
```diff
@@ -0,0 +1,238 @@
+    ![Figure 11: Rearranging the Sequence](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/sdh/release-stage-v2.gif)
+
+{% hint style="warning" %}
+The drag-and-drop feature is not meant for changing the sequence of applications within a stage. Use it solely for moving applications from one release stage to another.
```
This feels like, it's not meant to but it can? Should we say like it won't change in same stage but will across?

ashokdevtron replied:
No, @abhibhaw 
I rephrased it now, but the original intention was exactly that!
The drag-and-drop is not meant to but it can change the app sequence in same stage...So writing 'won't change in same stage' would be incorrect. I and @uxarya-d have discussed this earlier.

ashokdevtron replied:
This is rephrased line I am proposing "The drag-and-drop feature is designed specifically for moving applications between different release stages. It is not meant to alter the sequence of applications within the same stage."

abhibhaw commented on docs/user-guide/sdh/release-hub.md:
```diff
@@ -0,0 +1,238 @@
+
+### Putting a Release on Hold
+
+If a release is put on hold, none of the applications (in any release stage) can be deployed in that release. However, deploying from [Build & Deploy page](../deploying-application/README.md) of [Applications](../applications.md) or [Application Groups](../application-groups.md) will still be possible.
```
No need to refer apps and app grp...

ashokdevtron replied:
Removed

abhibhaw commented on docs/user-guide/sdh/release-hub.md:
```diff
@@ -0,0 +1,238 @@
+
+## Extras
+
+### Viewing Rollout Status
```
The rollout status has some grouping of statuses, when a tenant is in yet to trigger, when ongoing and failed. We need to cover that in doc

abhibhaw reviewed (changes requested):
I'm not sure if this is planned in next phase but few missing stuff:
1. How release stages dependency is goverened?
2. Talk about unmapped environment
3. Release notes and catalog of tenant and releases
4. Defining applications at release track (Guessing this will be in next phase of doc)
5. RBAC for deploying with release is not superadmin anymore (Guessing this will be in next phase of doc)
6. Release track Dashboard (Guessing this will be in next phase of doc)

ashokdevtron commented:
> I'm not sure if this is planned in next phase but few missing stuff:
> 
> 1. How release stages dependency is goverened?
> 2. Talk about unmapped environment
> 3. Release notes and catalog of tenant and releases
> 4. Defining applications at release track (Guessing this will be in next phase of doc)
> 5. RBAC for deploying with release is not superadmin anymore (Guessing this will be in next phase of doc)
> 6. Release track Dashboard (Guessing this will be in next phase of doc)

1. Need to discuss with you
2. Need to discuss with you
3. Added 👍 
4. Hmm. Didn't get it.
5. Updated 👍 
6. Need to discuss with you

siddhant-khisty commented on docs/user-guide/sdh/release-hub.md:
```diff
@@ -0,0 +1,265 @@
+
+{% hint style="warning" %}
+### Who Can Perform This Action?
+Users need to have super-admin permission to select images for selected applications.
+{% endhint %}
+
+1. Select a workflow available for your application. All the [images](../../reference/glossary.md#image) available in the selected workflow will appear.
+
+    ![Figure 12: Selecting Image from Specific Workflow](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/sdh/workflow-selection.gif)
+
+2. Click **SELECT** next to the image you wish to deploy from the list.
+
+3. Repeat the above steps for other applications you added in the release.
+
+    ![Figure 13: Repeating Steps for Other Applications](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/sdh/image-selections.gif)
+
+4. Click **Save**.
+
+{% hint style="info" %}
```
We should mention that only the built images will appear in the dashboard. If there are no images present, the user will need to perform the CI stage and build the application first to obtain the image.

ashokdevtron replied:
Will include this in the next iteration of this doc

ashokdevtron replied:
Change in plans...committed in this iteration itself 👍

siddhant-khisty commented on docs/user-guide/sdh/release-hub.md:
```diff
@@ -0,0 +1,265 @@
+
+    ![Figure 14: Adding Release Instructions](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/sdh/release-instructions-v2.gif)
+
+6. Before freezing the requirements, make sure the release order is correct, add applications if needed, and include environments in tenants (if not done already). Once you have finalized them, click **Lock Requirements**. 
```
```suggestion
6. Before locking the requirements, make sure the release order is correct, add applications if needed, and include environments in tenants (if not done already). Once you have finalized them, click **Lock Requirements**. 
```

ashokdevtron replied:
Will pick this up in the next iteration of this doc

ashokdevtron replied:
Fixed...committed in this iteration itself 👍

siddhant-khisty commented on docs/user-guide/sdh/tenants.md:
```diff
@@ -0,0 +1,78 @@
+
+This involves mapping customer's environments to the tenant installation so that your updates are deployed to the correct environments.
+
+1. Click **Map Environment**.
```
I think we should provide a link to the doc page for "Creating an Environment" as well before talking about the mapping process.

ashokdevtron replied:
Will pick this up in the next iteration of this doc

ashokdevtron replied:
Thanks, added it now.

abhibhaw commented on docs/user-guide/sdh/README.md:
```diff
@@ -0,0 +1,78 @@
+
+## Advantages
+
+Devtron's Software Distribution Hub goes beyond basic deployment by providing end-to-end release management. Deployments involving manual processes might be prone to human error. However, Software Distribution Hub automates the [rollout](#rollout) process by enforcing [requirements](#requirements) for each release, and not just for one environment but multiple tenant environments.
```
```suggestion
Devtron's Software Distribution Hub goes beyond basic deployment by providing end-to-end release management. Deployments involving manual processes might be prone to human error. However, Software Distribution Hub streamlines the [rollout](#rollout) process by enforcing [requirements](#requirements) for each release, and not just for one environment but multiple tenant environments.
```

abhibhaw commented on docs/user-guide/sdh/README.md:
```diff
@@ -0,0 +1,78 @@
+
+### Installations
+
+One installation represents one deployment of your software for a specific [tenant](#tenants). Each installation serves as a separate instance of the software, customized for different stages of use or separate teams within your organization. For example, your organization might have three installations: 
```
Na adding `environment` may confuse user, we purposely agreed to not use environment in product

abhibhaw commented on docs/user-guide/sdh/README.md:
```diff
@@ -0,0 +1,78 @@
+
+### Release Tracks
+
+A Release Track in Devtron is where you organize software releases, similar to a project or application. Each release within a Release Track is a unique version of your software. For example, think of Kubernetes as a release track, with "v1.28.8" as one of its releases. This helps in managing different versions and updates of a software project in a structured manner, ensuring that all versions a ...
```
```suggestion
A Release Track in Devtron is where you organize software releases. Each release within a Release Track is a unique version of your software. For example, think of Kubernetes as a release track, with "v1.28.8" as one of its releases. This helps in managing different versions and updates of a software project in a structured manner, ensuring that all versions are tracked and organized within their respective tracks.
```

abhibhaw commented on docs/user-guide/sdh/release-hub.md:
```diff
@@ -0,0 +1,265 @@
+
+{% hint style="info" %}
+### Importance of Release Stages
+By default, your selected applications will be set to release in one go. However, you can also release them in stages. In other words, you can decide which set of applications should be released first, subsequently which ones to release next, and the ones to release last. <br /><br /> For example, if you have a new login feature (backend) and a dashboard update (frontend), you might want to relea ...
```
```For example, if you have a new login feature (backend) and a dashboard update (frontend), you might want to release the login feature first to ensure users can log in before seeing the updated dashboard. ``` Some other example please, this seems off

ashokdevtron replied:
This time, used AI to generate a new example. Couldn't think of something better myself.
`For example, if you're adding a new payment system (backend) and an updated checkout page (frontend), you would release the payment system first to ensure payments can be processed correctly.`

abhibhaw commented on docs/user-guide/sdh/release-hub.md:
```diff
@@ -0,0 +1,265 @@
+An application can be deployed on the tenant in the next release stage only if other applications in the previous stage are deployed successfully on the given tenant.
+{% endhint %}
+
+Here we covered the process of performing a production installation on one tenant. Similarly, you can per
```
Similarly, you can per... Have we missed something here?

ashokdevtron replied:
Fixed

abhibhaw commented on docs/user-guide/sdh/release-hub.md:
```diff
@@ -0,0 +1,265 @@
+
+### Viewing Rollout Status
+
+You can view the status of installations under `Rollout Status`. Moreover, you can go to **Rollout History** tab to view the deployment history.
```
```suggestion
You can view the status of your release at a particular tenant under `Rollout Status`. Moreover, you can go to **Rollout History** tab to view the deployment history.
```

abhibhaw commented on docs/user-guide/sdh/release-hub.md:
```diff
@@ -0,0 +1,265 @@
+| Yet to Trigger    | Shows tenant installations for which rollout has not yet started  |
+| Ongoing           | Shows tenant installations for which deployment is currently in progress |
+| Failed            | Shows tenant installations where rollout has encountered issues because some deployment stage failed |
+| Completed         | Shows tenant installations where deployment has successfully finished and services are live   |
```
This needs some more clarification, but let's plan that in v2 of this doc

ashokdevtron replied:
Sure

abhibhaw commented on docs/user-guide/sdh/tenants.md:
```diff
@@ -0,0 +1,78 @@
+
+2. Enter a name in **Tenant display name** field, e.g., `flareup.xyz`
+
+3. Add a unique identifier to your tenant in **Tenant ID** field, e.g., `flareup123` 
```
This cannot be changed

abhibhaw replied:
Can only use alpha-numeric and `-`

ashokdevtron replied:
1. Agreed on the first suggestion...
2. No, let's not include field validations in the doc especially if they are already shown to the user as they type
<img width="1378" alt="image" src="https://github.com/user-attachments/assets/af2da662-8be8-49fe-9f7c-fde78980000c">

abhibhaw commented on docs/user-guide/sdh/tenants.md:
```diff
@@ -0,0 +1,78 @@
+
+    ![Figure 1: Adding a Tenant](https://devtron-public-asset.s3.us-east-2.amazonaws.com/images/sdh/tenants-tab.jpg)
+
+2. Enter a name in **Tenant display name** field, e.g., `flareup.xyz`
```
This can be renamed if required

abhibhaw commented on docs/user-guide/sdh/tenants.md:
```diff
@@ -0,0 +1,78 @@
+
+3. Add a unique identifier to your tenant in **Tenant ID** field, e.g., `flareup123` 
+
+4. (Optional) Add a description of the tenant.
```
Word limit: 300 characters I guess

ashokdevtron replied:
Tested. It's 350.
The user will likely not benefit it by reading it from the doc since this field validation will be shown on the UI if the user exceeds the limit. 
<img width="1441" alt="image" src="https://github.com/user-attachments/assets/2c098b63-bcaf-421a-ad6a-4543ef60bf1a">

abhibhaw reviewed (approved):
Suggested minor changes, can be picked up in next version of this doc too
